### DISCUSSION TITLE
Nice to have: consistent error propagation in main()

### DISCUSSION THREAD

**Discussion by dumblob** [vlang/v#11860] [Ideas]
This source:
```v
fn f()?{
  return error( 'my err' )
}
fn main() {
  f()?
}
```
yields:
```
V panic: optional not set (my err)
v hash: e9bbb7d
/tmp/v/a.7836897204962053226.tmp.c:5649: at v_panic: Backtrace
/tmp/v/a.7836897204962053226.tmp.c:5620: by panic_optional_not_set
/tmp/v/a.7836897204962053226.tmp.c:9392: by main__main
/tmp/v/a.7836897204962053226.tmp.c:9478: by main
```

But
```v
fn f()?{
  return error( 'my err' )
}
fn main() {
  f() or{ panic( err ) }
}
```
yields:
```
V panic: my err
v hash: e9bbb7d
/tmp/v/a.11159072267687342120.tmp.c:5649: at v_panic: Backtrace
/tmp/v/a.11159072267687342120.tmp.c:9393: by main__main
/tmp/v/a.11159072267687342120.tmp.c:9479: by main
```

I think it would be nice to have both behave identically (at least that's the semantics I'd expect). It'll be useful for proper error handling as tracked in https://github.com/vlang/v/issues/9367 .

**Comment by JalonSolov** [comment]
They are 2 different things, so I expect 2 different outputs.

In the first, you leave it to V to panic, and it tells you there was nothing in your code to handle it.

In the second, you explicitly did a panic, so it does as requested.

**Comment by dumblob** [comment]
Could you be more specific? The backtraces say the same thing except one being longer due to (IMHO unnecessary) a wrapper.

I know `main()` is a special case, but I can't fully agree, that I (the programmer) didn't handle the optional - actually vice versa - I **explicitly** (i.e. I did handle it) decided to "push it up" but because there is no more up on my system, it becomes a regular `panic()`.

Another thing is that the message `optional not set` is very misleading and I have no idea where (why) it comes from. The optional **is** set (as the parenthesized content `(my err)` says).

**Comment by JalonSolov** [comment]
The first one tells you:
```
/tmp/v/a.7836897204962053226.tmp.c:5620: by panic_optional_not_set
```
because you didn't handle the option yourself.

That line is missing in the second, because you did explicitly handle it yourself.

**Comment by dumblob** [comment]
Before I'll go into details, let's clarify one thing:

Is there any other situation except for the one I present in this issue when *V* can terminate with the `V panic: optional not set (...)` message? Please elaborate.

**Comment by dumblob** [comment]
I couldn't find any other place than `main()` to terminate with this special message.

Why is this very specific case then handled differently? Why is there this exception? Why maintaining superfluous code adding no value but causing confusion?

Note after the RFC https://github.com/vlang/rfcs/pull/7 from @danieldaeschle gets (hopefully) implemented, this will become even more confusing.
